csharp/src/Google.Protobuf/JsonFormatter.cs

```diff
@@ -248,6 +241,31 @@ private bool WriteMessageFields(TextWriter writer, IMessage message, bool assume
             return !first;
         }
 
+        /// <summary>
+        /// Determines whether or not a field value should be serialized according to the field,
+        /// its value in the message, and the settings of this formatter.
+        /// </summary>
+        private bool ShouldFormatFieldValue(IMessage message, FieldDescriptor field, object value)
+        {
+            bool formatDefaultValue = settings.FormatDefaultValues &&
```

Comment:
Can this be simply:

```
bool formatDefaultValue = settings.FormatDefaultValues && !field.HasPresence
```

Reply:
No. If FormatDefaultValues is set to true, then I'd expect a singular `int32` field which is either in a proto2 message or explicitly optional in a proto3 message to be formatted regardless of presence... although I can see arguments in favour of making field presence trump FormatDefaultValues.

I think we should probably decide that point, and then decide the implementation separately.